docs/deprecation-policy.md

```diff
@@ -0,0 +1,184 @@
+---
+assignees:
+- bgrant0607
+- lavalamp
+- thockin
+
+---
+
+# Kubernetes Deprecation Policy
+
+Kubernetes is a large system with many components and many contributors.  As
+with any such software, the feature set naturally evolves over time.  Sometimes
+we realize that something we have done in the past, whether that is some API, a
+flag, or even a whole feature, needs to be replaced or removed.  Kubernetes is
+committed to not breaking users, so we can’t just make breaking changes.  This
+is when deprecation is triggered.
+
+This document details the policies we hold on deprecation for various facets of
+the system.
+
+## Deprecating parts of the API
+
+Since Kubernetes is an API-driven system, it’s not surprising that the API
+would evolve as our understanding of the problem space evolves.  The Kubernetes
+API is actually a set of APIs, called “API groups”, and each API group is
+versioned independently.  [API versions](http://kubernetes.io/docs/api/)
+fall into 3 main tracks, each of which has different policies for deprecation:
+
+| Example  | Track                            |
+|----------|----------------------------------|
+| v1       | GA (generally available, stable) |
+| v1beta1  | Beta (pre-release)               |
+| v1alpha1 | Alpha (experimental)             |
+
+A given release of Kubernetes can support any number of API groups and any
+number of versions of each.
+
+The following rules govern the deprecation of elements of the API.  This
+includes:
+   * REST resources (aka API objects)
+   * Fields of REST resources
+   * Enumerated or constant values
+   * Component config structures
+
+These rules are enforced between official releases, not between
+arbitrary commits to master or release branches.
+
+**Rule #1: API elements may only be removed by incrementing the version of the
+API group.**
+
+Once an API element has been added to an API group at a particular version, it
+can not be removed from that version or have its behavior significantly
+changed, regardless of track.
+
+Note: For historical reasons, we have 2 “monolithic” API groups - “core” (no
+group name) and “extensions”.  We will incrementally move resources from these
+legacy API groups into more domain-specific API groups.
+
+**Rule #2: API objects must be able to round-trip between API versions in a given
```

Comment:
@thockin What's the policy for adding completely new features in a new API version? Example we have in mind is adding support for custom metrics in v2 HPA API.